### PR TITLE
Implement missing functions in mlx backend

### PR DESCRIPTION
**Pull request by Faisal-Alsrheed**
The `median` function has successfully passed all 30 test cases 

#### Test Outcome
```Python

======================================= test session starts ========================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.4.0
rootdir: /content
plugins: anyio-3.7.1
collected 30 items                                                                                 

test_median.py ..............................                                                [100%]

======================================== 30 passed in 0.35s ========================================
```

Here’s a summarized overview of the tests performed on the `median` function:

- **Basic Functionality**: Tested the function's ability to calculate the median of single-dimensional arrays without altering the dimensions or using the `keepdims` parameter.
- **Multidimensional Support**: Validated calculations for two-dimensional and three-dimensional arrays, both with and without keeping the dimensions (using `keepdims`).
- **Negative and Multiple Axes**: Assessed the function’s correctness over negative axes and multiple axes simultaneously, ensuring accuracy across different array configurations.
- **Edge Cases**: Included tests for arrays with single elements, identical elements, and very high dimensionality, ensuring the function handles special cases correctly.
- **Large Arrays**: Verified the performance on large arrays to ensure computational efficiency and accuracy.
- **Non-Integer and Random Values**: Tested with non-integer values and arrays filled with random numbers to confirm consistency across various data types.

```Python

import numpy as np
import pytest
from median import median

def test_no_axis_no_keepdims():
    assert np.allclose(median(np.array([1, 2, 3, 4, 5]), axis=None, keepdims=False),
                       np.median(np.array([1, 2, 3, 4, 5]), axis=None, keepdims=False))

def test_1d_array_axis_0():
    assert np.allclose(median(np.array([1, 2, 3, 4, 5]), axis=0, keepdims=False),
                       np.median(np.array([1, 2, 3, 4, 5]), axis=0, keepdims=False))

def test_1d_array_axis_0_keepdims():
    assert np.allclose(median(np.array([1, 2, 3, 4, 5]), axis=0, keepdims=True),
                       np.median(np.array([1, 2, 3, 4, 5]), axis=0, keepdims=True))

def test_2d_array_axis_0_no_keepdims():
    assert np.allclose(median([[1, 2], [3, 4]], axis=0, keepdims=False),
                       np.median([[1, 2], [3, 4]], axis=0, keepdims=False))

def test_2d_array_axis_0_keepdims():
    assert np.allclose(median([[1, 2], [3, 4]], axis=0, keepdims=True),
                       np.median([[1, 2], [3, 4]], axis=0, keepdims=True))

def test_2d_array_axis_1_no_keepdims():
    assert np.allclose(median([[1, 2], [3, 4]], axis=1, keepdims=False),
                       np.median([[1, 2], [3, 4]], axis=1, keepdims=False))

def test_2d_array_axis_1_keepdims():
    assert np.allclose(median([[1, 2], [3, 4]], axis=1, keepdims=True),
                       np.median([[1, 2], [3, 4]], axis=1, keepdims=True))

def test_3d_array_axis_0_no_keepdims():
    assert np.allclose(median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=0, keepdims=False),
                       np.median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=0, keepdims=False))

def test_3d_array_axis_0_keepdims():
    assert np.allclose(median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=0, keepdims=True),
                       np.median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=0, keepdims=True))

def test_3d_array_axis_1_keepdims():
    assert np.allclose(median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=1, keepdims=True),
                       np.median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=1, keepdims=True))

def test_3d_array_axis_2_keepdims():
    assert np.allclose(median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=2, keepdims=True),
                       np.median([[[1, 2], [3, 4]], [[5, 6], [7, 8]]], axis=2, keepdims=True))

def test_2d_array_axis_negative_1():
    assert np.allclose(median([[1, 2], [3, 4]], axis=-1, keepdims=False),
                       np.median([[1, 2], [3, 4]], axis=-1, keepdims=False))

def test_2d_array_axis_negative_2_keepdims():
    assert np.allclose(median([[1, 2], [3, 4]], axis=-2, keepdims=True),
                       np.median([[1, 2], [3, 4]], axis=-2, keepdims=True))

def test_single_element_array():
    assert np.allclose(median([42], axis=0), np.median([42], axis=0))

def test_all_identical_elements():
    assert np.allclose(median([7, 7, 7, 7], axis=0), np.median([7, 7, 7, 7], axis=0))

def test_large_array():
    large_array = np.ones(10000) * 5
    assert np.allclose(median(large_array, axis=0), np.median(large_array, axis=0))

def test_non_integer_values():
    assert np.allclose(median([1.5, 2.5, 3.5], axis=0), np.median([1.5, 2.5, 3.5], axis=0))


def test_negative_axis_wrapping():
    array = np.random.rand(10, 20)
    assert np.allclose(median(array, axis=-1), np.median(array, axis=-1))


def test_keepdims_consistency():
    array = np.random.rand(5, 10)
    result = median(array, axis=1, keepdims=True)
    assert result.shape == (5, 1)


# Test for multi-axis support in 2D arrays
def test_2d_array_multi_axis():
    array = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
    expected = np.median(array, axis=(0, 1))
    result = median(array, axis=(0, 1))
    assert np.allclose(result, expected), "Failed: 2D array multi-axis median calculation"

# Test for multi-axis support in 3D arrays
def test_3d_array_multi_axis():
    array = np.random.rand(2, 3, 4)
    expected = np.median(array, axis=(1, 2))
    result = median(array, axis=(1, 2))
    assert np.allclose(result, expected), "Failed: 3D array multi-axis median calculation"

# Test with keepdims=True and multiple axes
def test_multi_axis_keepdims():
    array = np.random.rand(2, 3, 4, 5)
    expected = np.median(array, axis=(0, 2), keepdims=True)
    result = median(array, axis=(0, 2), keepdims=True)    
    assert np.allclose(result, expected), "Failed: Median calculation with keepdims on multiple axes"

# Test with negative axes
def test_negative_axes():
    array = np.random.rand(4, 5, 6)
    expected = np.median(array, axis=(-2, -1))
    result = median(array, axis=(-2, -1))
    assert np.allclose(result, expected), "Failed: Negative axes median calculation"


def test_multi_axis_keepdims_3d():
    array = np.random.rand(2, 3, 4)
    expected = np.median(array, axis=(0, 2), keepdims=True)
    result = median(array, axis=(0, 2), keepdims=True)
    assert np.allclose(result, expected), "Failed: Median calculation with keepdims on multiple axes in a 3D array"

def test_multi_axis_keepdims_5d():
    array = np.random.rand(2, 3, 4, 5, 6)
    expected = np.median(array, axis=(1, 3), keepdims=True)
    result = median(array, axis=(1, 3), keepdims=True)
    assert np.allclose(result, expected), "Failed: Median calculation with keepdims on multiple axes in a 5D array"

def test_multi_axis_keepdims_all_axes():
    array = np.random.rand(2, 3, 4)
    expected = np.median(array, axis=(0, 1, 2), keepdims=True)
    result = median(array, axis=(0, 1, 2), keepdims=True)
    assert np.allclose(result, expected), "Failed: Median calculation with keepdims on all axes"

def test_multi_axis_keepdims_negative_axes():
    array = np.random.rand(2, 3, 4, 5)
    expected = np.median(array, axis=(-2, -1), keepdims=True)
    result = median(array, axis=(-2, -1), keepdims=True)
    assert np.allclose(result, expected), "Failed: Median calculation with keepdims on negative axes"

def test_very_high_dimensionality():
    array = np.random.rand(2, 2, 2, 2, 2, 2, 2, 2)
    expected = np.median(array, axis=(0, 1))
    result = median(array, axis=(0, 1))
    assert np.allclose(result, expected), "Failed: High dimensionality median calculation"

def test_all_identical_elements_keepdims():
    array = np.array([42, 42, 42, 42])
    result = median(array, axis=0, keepdims=True)
    expected_shape = (1,)
    assert result.shape == expected_shape, f"Expected shape {expected_shape}, got {result.shape}"


def test_vector_median_keepdims():
    array = np.array([10, 20, 30])
    result = median(array, axis=0, keepdims=True)
    expected = np.array([20])
    assert np.array_equal(result, expected), f"Expected {expected}, got {result}"

```

### CODE REVIEW

File: keras/src/backend/mlx/numpy.py [hunk 1]
```diff
@@ -579,21 +579,33 @@ def maximum(x1, x2):
...
-    x = convert_to_tensor(x)
-    x_sorted = mx.sort(x, axis=axis)
-    axis_size = x_sorted.shape[axis]
-    medians = mx.take(
-        x_sorted, indices=mx.array([(axis_size // 2) - 1]), axis=axis
-    )
-    if not keepdims:
-        medians = mx.squeeze(medians, axis=axis)
-    return medians
+def median(x, axis=None, keepdims=False):
```
Comment (fchollet) [inline keras/src/backend/mlx/numpy.py:582]: I think this implementation will not support a non-scalar axis or negative axes. Can you check if this was tested? The previous implementation should support it (from reading the code).

**Comment by Faisal-Alsrheed** [comment]
Done.
@fchollet Thank you for your feedback.

**Comment by Faisal-Alsrheed** [comment]
What I've learned from this experience is that the best way to use a large language model (LLM) to write a function is to first create comprehensive tests with pytest and then Keep sharing the full results with the LLM.

Sometimes, if it’s a complex problem, I need to use multiple models until I get it right. 

Test-driven development is fun :)

**Review by fchollet — APPROVED** [review]
Nice, thank you!
